bot/src/discord_commands/schedule.js

```diff
@@ -0,0 +1,165 @@
+const { Permissions } = require('monochrome-bot');
+const randomHelper = require('./random.js');
+
+let intervals = [];
```

Comment:
Maybe this should be `const intervalsForServerId = {};` and each server has it's own list of intervals. Otherwise it seems like the clear command will clear all intervals across all servers, and the list command will show all intervals across all servers? Need to keep things siloed per server.

So then you would get the intervals for a server with `const intervals = intervalsForServerId[serverId] || [];`

Reply:
I assumed that every server run an independent instance of the bot. So this way you could get the list of intervals of any server? Maybe we could store the intervals in the bot settings? In a new subcategory? I'll look into that.

Reply:
We could do that. It would look similar to the shiritori forever setting [here](https://github.com/ernespa/kotoba/blob/dcf16cd8040474c3e27c708b24e263a83f1f7bc0/bot/src/bot_settings.js#L314).

Reply:
I've seen that with monochrome.getSettings() I can access to the bot settings, but still don't know where are those settings stored or if we can change them programmatically. My idea was to use the settings functionality for storing data but not necessarily making it a new public setting. But if there's an easier way to store those commands and execute all of them when the bot resets, that would be perfect.
I should edit the code so the schedule is not dependant on the moment you run it (it isn't, but if you set it to run every 5 hours starting at 12:00, after 24 hours it won't run at 12:00 but at 13:00, so if the bot runs the command after rebooting, it would mistakenly start at 12.00 (we could just limit the possible frequencies to divisors of 24)

Reply:
The monochrome settings API is only useful for public settings. To store arbitrary data, you'd have to interact with the database directly. I can tell you about how it works if you want to try going down that route. Kotoba uses MongoDB and mongoose ODM.

To develop features that use the database, a MongoDB server is required to be running on your computer. The easiest way to start one is with Docker: `docker run -p 27017:27017 mongo:4.2`. Alternatively if not using Docker, MongoDB can be installed [here](https://www.mongodb.com/try/download/community). Kotoba uses version 4.2 but the latest version is probably fine.

Then you would need to write mongoose schemas for the new data, that would look something like this:

```js
const mongoose = require('mongoose');

const ChannelWordScheduleSchema = new mongoose.Schema({
  _id: { type: String, required: true }, // channel ID
  frequency: { type: Number, required: true },
  start: { type: Number, required: true },
  level: { type: String, required: false },
});

const ServerWordScheduleschema = new mongoose.Schema({
  _id: { type: String, required: true }, // server ID
  channelSchedules: { type: [ChannelWordScheduleSchema], required: true },
});

function create(connection) {
  return connection.model('ServerSchedule', ServerWordScheduleschema);
}

module.exports = create;
```

And would go in a new file [here](https://github.com/mistval/kotoba/tree/master/node-common/mongoose_models) and would also have to be exported from the models index file [here](https://github.com/mistval/kotoba/blob/master/node-common/mongoose_models/index.js).

Then you can require the model in the command like:

```js
const dbConnection = require('kotoba-node-common').database.connection;
const ServerWordSchedules = require('kotoba-node-common').models.createServerWordSchedulesModel(dbConnection);
```

Then I think it's fairly straightforward. You can get all schedules for a server by doing:

```js
const serverSchedules = await ServerWordSchedules.findById(serverId);
```

And you can add a new one by doing:

```js
let serverSchedules = await ServerWordSchedules.findById(serverId);

if (!serverSchedules) {
  // no schedules found in the database for this server.
  // gotta create a brand new DB entry
  serverSchedules = new ServerWordSchedules({
    _id: serverId,
    channelSchedules: [],
  });
}

serverSchedules.channelSchedules.push({
  _id: channelId,
  frequency: frequency,
  start: start,
  level: level,
});

await serverSchedules.save(); // put it into the database
```

One thing is that you can only store JSON (technically BSON) serializable data in the database, so you couldn't store for example the `msg` object. So you would have to refactor things a bit to rely on having the channel, not the message. That could be slightly tricky. I can do that part, if you do want to try this route.

Reply:
Thank you, I think I understood everything. I'll give it a shot in a couple of days.

Reply:
Ok, so the database part is working fine, I can create entries and list them, If we limit it to one interval per channel, then the stop command doesn't need a parameter and just stops that channel's interval (and the clear command still removes all of them).

Now I wonder if we should have just a lot of Timeouts running or if it's better to have a single Interval that once per minute checks the list and runs the corresponding commands.

What do you think?

Reply:
I added a status property so you can pause an interval without deleting and resuming it later.

Reply:
Ah, now I see that we need the msg object to call the _random_ command, we could fake it at runtime, only with the properties we need...

Reply:
A bunch of timers vs one interval, I think either would be fine. I would probably lean towards one interval because then you can query the database to get all the schedules that are due. If we did that we would probably want to redesign the mongoose schemas. Maybe it would just be like:

```js
const WordScheduleSchema = new mongoose.Schema({
  _id: { type: String, required: true }, // channel ID
  serverId: { type: String, required: true, index: true }, 
  frequency: { type: Number, required: true, index: true },
  start: { type: Date, required: true, index: true },
  lastSent: { type: Date, required: false, index: true },
  level: { type: String, required: false },
});
```

Then you could get all currently due schedules like:

```js
const dueSchedules = await WordScheduleModel.findMany({
  start: { $lt: new Date() },
  $expr: { $lt: ['$lastSent', { $subtract: [new Date(), '$frequency']  }] }
});

for (const schedule of dueSchedules) {
  // send the message
  ...
  schedule.lastSent = new Date();
  await schedule.save();
}
```

Regarding needing the message, I will refactor it this weekend so it only needs the channel.

Reply:
I refactored the random word stuff so you can give it a channel instead of a message (https://github.com/mistval/kotoba/commit/45bad23579b7eec30f832b35d42a13853fc15c02). So now if you have the server ID and channel ID you can use it like:

```js
const showRandomWord = require('./../discord/show_random_word.js');

const channel = monochrome.getErisBot().guilds.get(serverId).channels.get(channelId);
await showRandomWord(level, channel, monochrome);
```
Also it's possible the channel or server is deleted or Kotoba no longer has permission to send messages in it, so that will need to be handled gracefully.

Reply:
If a user runs the stop option, should we delete that entry from the database or would it be safer to just tag it as deleted, and clean the list when the bot restarts?

Reply:
Deleting it from the database would be best. To delete all schedules for a server (assuming the above schema) would be:

```js
await WordScheduleModel.deleteMany({ serverId: serverId });
```

To delete the schedule for a channel would be:

```js
await WordScheduleModel.deleteOne({ _id: channelId });
```

Reply:
I think it's done, I don't know much about how should we handle exceptions and permissions. Also when I run the command 'k!help schedule' I get "Cooldown: 0 seconds', which I don't know what does it mean, and 'Required permissions: None', which should be Server Admin or something like that.

I've seen that, since we're not using the message object, there's no pagination in the results, does that mean that we're missing the next pages or it's not paginated at all?

I think we could add also the k!strokeorder and the k!examples commands to the mix. Or at least set a bigger font to the word itself.

Reply:
I wasn't sure what would be best to do about the pagination. If we added the pagination back we could either do:

1. Only the user that created the schedule can use the pagination buttons.
2. Anyone can use the pagination buttons (I could see that being a bit chaotic in big servers).

Or instead of just sending one message, we could send several. Maybe the first definitions page, the first examples page, and the first kanji page. What do you think?

Reply:
It depends on how the random search work. For example, I got the word 誰 (dare) and it showed the words 誰, 誰か and 誰にも, but can the random search return 誰か directly? I think for a word of the day is enough with one of the results (誰 returns 46 results on jisho.org, I get why we should't show all of them in the same message).

If we can show just one word, then I don't see a problem in showing the strokeorder of all of its kanjis, and the first page of examples.

We could also add a few lines at the end showing useful commands like 'Type 'k!examples 誰' to see more examples', so people can easily copy and paste that.

Reply:
Would you want to design a new message for this, instead of using what the random command sends? Maybe send the first definition, some examples, maybe some other stuff in one message, and send some kanji info in subsequent messages or something. You can get the raw JSON from Jisho like:

```js
const JishoApi = require('unofficial-jisho-api');
const globals = require('./../common/globals.js');

const jisho = new JishoApi();

const word = await globals.resourceDatabase.getRandomWord(level);
const definitions = await jisho.searchForPhrase(word);
const kanji = await jisho.searchForKanji(kanji);
const examples = await jisho.searchForExamples(word);
```

Reply:
I added two new optional parameters to showRandomWord: showStrokeOrder and showExamples, and added this at the end of the function:

```js
  if (showStrokeOrder || showExamples) {
    firstPage.embed.fields = [firstPage.embed.fields[0]]; // discards the rest of the results
    await channel.createMessage(firstPage);
  }
  if (showStrokeOrder) {
    for (let i = 0; i < word.length; i++) {
      const stroke = await createStrokeOrderSearchPage(word[i]); // shows the strokes of every character (would be nice to ignore kanas)
      if (!showExamples && i === word.length - 1) {
        return channel.createMessage(stroke);
      } else {
        await channel.createMessage(stroke);
      }
    }
  }
  if (showExamples) {
    let example = await createExampleSearchPages(word);
    example = example[0]; // discards the other pages
    example.embed.title = example.embed.title.split(' (')[0]; // removes the '(Page 1 of X)' from the title
    return channel.createMessage(example);
  }
  return channel.createMessage(firstPage);
```

It's just an idea, what do you think?

Reply:
Some improvements, pretty happy with the result right now.

```js
  if (showStrokeOrder || showExamples) {
    if (firstPage.embed.fields.length > 1) {
      firstPage.embed.title += ' (click here for more results)';
    }
    firstPage.embed.fields = [firstPage.embed.fields[0]]; // discards the rest of the results
    await channel.createMessage(firstPage);
  }
  if (showStrokeOrder) {
    let regexp = /[\u3040-\u30FF]/; // Hiragana and katakana charset
    for (let i = 0; i < word.length; i++) {
      const isKana = regexp.test(word[i]);
      if (!isKana) {
        const stroke = await createStrokeOrderSearchPage(word[i]); // shows the strokes of every kanji
        await channel.createMessage(stroke);
      }
    }
    if (!showExamples) return null;
  }
  if (showExamples) {
    let example = await createExampleSearchPages(word);
    if (example.length > 1) {
      example[0].embed.title = example[0].embed.title.split(' (')[0]; // removes the '(Page 1 of X)' from the title
      example[0].embed.title += ' (click here for more examples)';
    }
    example = example[0]; // discards the other pages
    return channel.createMessage(example);
  }
  return channel.createMessage(firstPage);
```

Reply:
This seems good. To get all the kanji in word you can also do `const kanji = word.match(/([\u4e00-\u9faf])/g);`

Reply:
One last thing, I was missing the definition of the individual kanjis, so I decided to change it to the 'kanji search' command and, with a new 'strokes' parameter, replace the thumbnail with the gif, and the examples part with the stroke order image. I don't like to mess the other functions that much but this would be my preferred result.

```js
  if (!strokes) {
    const examplesStr = getExamplesString(kanjiData);
    if (examplesStr) {
      embedFields.push({ name: 'Examples', inline: false, value: examplesStr });
    }
  }

  const unicodeString = kanjiData.query.codePointAt(0).toString(16);
  let gifFound = false;
  if (strokes) {
    const fileNameGif = `${unicodeString}.gif`;
    const animationUri = `https://raw.githubusercontent.com/mistval/kanji_images/master/gifs/${fileNameGif}`;
    gifFound = await axios.get(animationUri);
  }
  const fileName = gifFound ? `${unicodeString}.gif` : `${unicodeString}.png`;
  const thumbnailInfo = gifFound ? { url: `https://raw.githubusercontent.com/mistval/kanji_images/master/gifs/${fileName}` } : { url: `https://raw.githubusercontent.com/mistval/kanji_images/master/pngs/${fileName}` };

  return {
    embed: {
      title: kanjiData.query,
      description: getDescriptionString(kanjiData),
      url: kanjiData.uri,
      fields: embedFields,
      image: strokes ? { url: kanjiData.strokeOrderDiagramUri } : null,
      thumbnail: thumbnailInfo,
      color: constants.EMBED_NEUTRAL_COLOR,
      footer: strokes ? null : {
        text: `Wanna see detailed stroke information for this Kanji? Try '${prefix}so ${kanjiData.query}'`,
        icon_url: constants.FOOTER_ICON_URI,
      },
    },
  };
```

Reply:
This is how it looks:
![image](https://user-images.githubusercontent.com/28624672/99910151-c2fa0000-2cec-11eb-9bd6-7f7a93c6b3e3.png)

Reply:
That seems good. Just one thing: the [kanji images repo](https://github.com/mistval/kanji_images) has gifs and pngs for the same list of kanji. So if the gif isn't there, the png won't be there either. So I think you can skip the axios request and just use the gif as the thumbnail, and if the gif isn't available, I believe Discord will just not render it and there won't be a thumbnail.

Reply:
Ah, ok, I did it that way because the axios command was being used in the stroke order command, but not in the kanji search one.

Reply:
The stroke order command does that for a special case. If it does not find the kanji on Jisho, then it has to know if the gif exists or not so it can decide whether to send the gif alone, or send a "no results found" message.